enhancements/catalog-data-transfer-reduction.md

```diff
@@ -0,0 +1,450 @@
+      define dependencies := []
+      foreach gvk in bundle.props.requiredGVKs:
+        dependencies.push(input_new.getLatestForGVK(gvk))
+      end
+      foreach package in bundle.props.requiredPackages:
+        dependencies.push(input_new.getLatestForPackage(package))
+      end
+      foreach packageRange in bundle.props.requiredPackageRanges:
+        dependencies.push(input_new.getLatestForPackageRange(packageRange))
+      end
+
+      define deps_by_channel := groupByChannel(dependencies)
+      foreach channel, deps := deps_by_channel:
+        define first := deps.least_version
+        define last := deps.greatest_version
+        output.push(diffChannel(channel, first, last))
+        output.push(first)
+        output.push(last)
+      end
+    end
+  end
+
+end
+```
+
+#### Diff inputs and catalog snapshotting
+
+A DC makes sense as output‘s type because they represent all bundle metadata
+required to build an index image in one unified format,
+and `opm render` can consume any input format to produce
+a normalized declarative config which can be worked with internally.
+
+A declarative config makes less sense as `input_expand` since diff only
+requires catalog object identifiers to keep or remove their identifying
+objects from a catalog. More pertinent is the disconnected catalog's size:
+transferring an entire catalog from disconnected to connected is equally
+as un-scalable as connected to disconnected. Instead, a simple *catalog snapshot*,
+a list of packages containing channel and bundle elements, is easy
+to both produce from in-cluster subscriptions and manipulate by hand:
+
+```yaml
+packages:
+- name: foo
+  channels:
+    - name: preview
+      head: foo.v0.1.0-beta.2
+      versions:
+      - 0.1.0-beta.0
+      - 0.1.0-beta.1
+      - 0.1.0-beta.2
+    - name: stable
+      head: foo.v0.1.0
+      versions:
+      - 0.1.0
+- name: bar
+- name: baz
+  channels:
+...
+```
+
+Head information is only required if performing a FromHeads diff.
+
+Optionally a DC describing the full disconnected catalog state can be provided instead of a snapshot,
+and can be transformed into the snapshot format internally.
+
+A standalone tool or new `opm` command will be written to build a catalog snapshot
+from a cluster by making [`ListBundles()`][listbundles] gRPC calls to all registries running.
+
+#### Merging catalogs
+
+Now that the diff `output` has been created, some mirroring command like `oc adm catalog mirror`
+will send the `output` (DC) and all referenced images to a disk mount.
+The mount is then transferred via meatspace to the disconnected side,
+and images are again mirrored from the mount to a registry.
+
+If there is no existing catalog in the cluster (fresh cluster),
+then a CatalogSource and set of Subscriptions are created and the process is complete.
+
+If one or more catalogs do exist in the cluster, then a new CatalogSource
+must be created for `output`. This is already possible to do.
```

Comment:
That's possible to do with a few commands. IIRC I had this thought initially but didn't include it because it was a non-trivial amount of extra work for cluster admins. However I think this workflow could be fully automated on the disconnected side. I will add it to this section.